### ISSUE TITLE
Decoding the database file

### ISSUE THREAD

**Issue by henrybetts** [Deep-Symmetry/dysentery#10]
Hi,

Thanks for your work on documenting this. I think I will find it very useful in the near future.

I have managed to decode the database file that rekordbox generates (or at least the .pdb file that it saves to USB devices; the .edb file found in the user directory is slightly different). This file contains the metadata for each track; name, id etc... as well as the path to the associated .DAT file.

This means that it would be possible to do a fully offline analysis of tracks. Do you think this would be useful? Obviously, the ideal situation would be to request metadata directly from the players, without limitations - has anyone been able to test whether the Nexus 2's accept requests from any device?

Thanks

**Comment by brunchboy** [comment]
Henry, that is fantastic! I would love to learn what you have discovered, I would like very much to add USB database reading capabilities to beat-link, for people running shows to be able to quickly create a metadata archive from the DJ’s memory stick, when they are going to be using four players. I don’t have access to nxs2 devices to test this, but I have not heard that they do any better.

Can you please share what you have discovered? I would very much like to add it to the analysis document and implement it.

**Comment by henrybetts** [comment]
Cool, no problem. I am pretty busy this week, but at the weekend I should have some time to write up and post what I have learned so far.

**Comment by brunchboy** [comment]
That would be wonderful, thanks so much!

**Comment by brunchboy** [comment]
@henrybetts I really hope you will find time soon to share what you have learned, I can’t wait to build it into Beat Link Trigger, as I regularly hear from people who would like to be able to work reliably with a full set of players on the network.

**Comment by henrybetts** [comment]
@brunchboy Sorry it took so long to get back to you. I have started putting up some documentation here: https://github.com/henrybetts/Rekordbox-Decoding.

As you will see, there are still lots of gaps that need filling in, and a few things that need confirming. I will continue to update - I think I've worked out most of the Track, Artist and Album tables (there are others, but I think these three are what most people will be interested in). I have done a quick parsing example in C, which I will probably continue to evolve as well.

I hope it makes some sense. If you open a pdb file in a hex editor, you'll be able to see some of what I've described.  Feel free to ask any questions, or update if you discover anything that I've missed!

**Comment by flesniak** [comment]
I wrote a functional pdb parser in python. In my project, it is preferred over a dbserver connection due to the rendering issues when using player numbers > 4. It simply downloads the database using nfs and parses the database locally to answer database queries. It also contains a parser for dat and ext files in the usbanlz folder. I tested it on some events and the parser seems so work with every export.pdb the DJs brought along so far. Unfortunately, the code is a bit slow on large pdbs, it takes ~5 seconds for a 10MB database to load on a moderate laptop.

Its written using [python construct](https://construct.readthedocs.io/en/latest/), thus the definitions should be pretty easy to read from the code. I'll have a look at @henrybetts documentation to fix some unknown values in my own parser (i.e. i missed the KUVO stuff).

Code here: https://github.com/flesniak/python-prodj-link/tree/master/pdblib

**Comment by EvanPurkhiser** [comment]
Wow this is some absolutely amazing work. Looking forward to porting things over to my go library! Also awesome to see someone else who’s working on tooling in the space @flesniak.

P.S. @flesniak I was surprised to see Hardcore in your readme screenshots, let alone Getty! We should chat :)

**Comment by brunchboy** [comment]
Wow, yes, thanks so very much for figuring this stuff out and sharing it with us! I can’t wait to find some time to explore the details.

**Comment by brunchboy** [comment]
I’m finally working on porting this stuff. Thanks so much, @henrybetts and @flesniak for sharing your discoveries, I hope you will be able to answer questions if I get stuck. I am porting it to [Kaitai Struct](http://kaitai.io) so parsers for a variety of languages will be available. Their Web IDE is proving an amazing help in validating pieces as I implement them!

**Comment by brunchboy** [comment]
@flesniak if there is any way you could explain the structure of the reverse index at the end of the database pages in words (perhaps with some diagrams), the way @henrybetts did, it would be wonderful. I don’t understand that complicated Construct code, nor the python syntax (which you use even in some of your comments).

**Comment by flesniak** [comment]
Yes, the Reverse Index implementation is pretty messy. At the end of each page, there are two 16bit integers which can be used to disable entries. They are _prepended_ by a 16 bit integer for each entry in that page which corresponds to the start offset of that entry in bytes. So for example, a Page of three entries Could end with 80 48 16 255 255 (All numbers 16 bit wide), which means that entry 1 starts 16 bytes after the Page Header, entry 2 48 bytes after and entry 3 80 bytes after the Header. 255 255 is a bitmask saying every entry is valid (a value of 0 would disable the entry matching the bit Index).

Summed up, pages end with an array of N+2 int16s, N being the numbers of entries in that page. Entry x offset in bytes after Page Header is at Position N-x. 

Am 2. November 2018 04:56:18 MEZ schrieb James Elliott <notifications@github.com>:

**Comment by brunchboy** [comment]
Thanks for the confirmation @flesniak. I ended up staying up way too late last night and figuring some of that out, and the process of implementing a Kaitai version is coming along (one of their lead developers has gotten interested, which is helping a great deal 😀). It looks like the second bit mask, `entry_enabled_override` is not currently used in your code, or did I miss that?

**Comment by brunchboy** [comment]
Say, @flesniak, I think we can avoid that long time parsing large PDB files that you mentioned in January, if we figure out how to read the index tables. Then there would be no need to ever parse the entire file, simply look through the correct index file for the desired record, and load that page and row directly.

**Comment by brunchboy** [comment]
Thanks again, everyone! This has been implemented as a new project, Crate Digger, and is being used in a preview release of Beat Link Trigger. So awesome!

**Comment by henrybetts** [comment]
Wow, looks like you've done a great job! I'm sorry, I ran out of time and never really got a chance to return to the project.

That web IDE looks very interesting - I'll bookmark that.

**Comment by brunchboy** [comment]
I know how that goes, it took me many months to find time to get into this myself. I also still need to write the majority of the analysis article in Crate Digger that explains the details we have figured out about the packets, for people who don’t want to have to learn how to read `.ksy` files. But there is a Beat Link Trigger feature I promised an artist that I want to get back to and finish first. Always too much to do! 😂
